### PR TITLE
Gradient circuits 3/n: implementation for ParameterShift

### LINKED ISSUE #515: Implement get_gradient_circuits for ParameterShift

### PR DESCRIPTION
**Pull request by zaqqwerty**
Resolves #515 .  During implementation I discovered a more efficient way to represent the gradient structure: separating the old `batch_mapper` into a weights tensor and an index tensor.  This way there is no longer any zero padding necessary.

### CODE REVIEW

**Comment by MichaelBroughton** [comment]
> This way there is no longer any zero padding necessary.

@jaeyoo as the person who originally worked on this differentiator would you be able to check this out and see if everything still works as intended ?

**Comment by jaeyoo** [comment]
> > This way there is no longer any zero padding necessary.
> 
> @jaeyoo as the person who originally worked on this differentiator would you be able to check this out and see if everything still works as intended ?
> I think you and I should both review this one.

Yes I am starting review

**Review by jaeyoo — DISMISSED** [review]
LGTM! Thank you so much Tony! It's great work!

File: tensorflow_quantum/python/differentiators/parameter_shift.py [hunk 1]
```diff
@@ -56,12 +56,55 @@ class ParameterShift(differentiator.Differentiator):
...
+        batch_weights = tf.reshape(
+            tf.transpose(weights, [1, 0, 2, 3]),
+            [n_programs, n_symbols, n_param_gates * n_shifts])
+        shifts = tf.reshape(tf.transpose(shifts, [1, 0, 2, 3]),
+                            [n_programs, m_tile, 1])
+
+        # Append impurity symbol into symbol name
+        new_symbol_names = tf.concat([
+            symbol_names,
+            tf.constant([parameter_shift_util._PARAMETER_IMPURITY_NAME])
```
[the remark this answers is not in the document]
Comment (zaqqwerty) [reply tensorflow_quantum/python/differentiators/parameter_shift.py:91]: Seems there was a test flake, I'll update this and repush

[the remark this answers is not in the document]
Comment (zaqqwerty) [reply tensorflow_quantum/python/differentiators/parameter_shift.py]: Done

**Review by jaeyoo — APPROVED** [review]
LGTM

**Pull request merged**